File: examples/python/sign_rpm.py
"""Sign an RPM with a PGP private key."""

import argparse

from rpm_rs import Package, Signer

parser = argparse.ArgumentParser(description="Sign an RPM with a PGP private key")
parser.add_argument("rpm", help="Path to the RPM file")
parser.add_argument(
    "-k", "--key", required=True, help="Path to the private key file (ASCII-armored)"
)
parser.add_argument("-o", "--output", help="Output path (default: overwrite input)")
parser.add_argument("-p", "--passphrase", help="Key passphrase (if protected)")
args = parser.parse_args()

signer = Signer.from_file(args.key)
if args.passphrase:
    signer = signer.with_key_passphrase(args.passphrase)

output = args.output or args.rpm

if output == args.rpm:
    try:
        Package.resign_in_place(args.rpm, signer)
    except RuntimeError:
        pkg = Package.open(args.rpm)
        pkg.sign(signer)
        pkg.write_file(output)
else:
    pkg = Package.open(args.rpm)
    pkg.sign(signer)
    pkg.write_file(output)

print(f"Signed: {output}")

File: examples/sign-rpm.rs
use clap::Parser;
use rpm::{Package, signature::pgp::Signer};
use std::path::PathBuf;

/// Sign an RPM with a PGP private key.
#[derive(Parser)]
#[command(version)]
struct Args {
    /// Path to the RPM file
    rpm: PathBuf,

    /// Path to the private key file (ASCII-armored)
    #[arg(short, long)]
    key: PathBuf,

    /// Output path (defaults to overwriting the input file)
    #[arg(short, long)]
    output: Option<PathBuf>,

    /// Key passphrase (if the key is protected)
    #[arg(short, long)]
    passphrase: Option<String>,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();

    let mut signer = Signer::from_asc_file(&args.key)?;
    if let Some(passphrase) = args.passphrase {
        signer = signer.with_key_passphrase(passphrase);
    }

    let output = args.output.as_ref().unwrap_or(&args.rpm);

    if output == &args.rpm {
        if Package::resign_in_place(&args.rpm, signer.clone()).is_err() {
            let mut pkg = Package::open(&args.rpm)?;
            pkg.sign(signer)?;
            pkg.write_file(output)?;
        }
    } else {
        let mut pkg = Package::open(&args.rpm)?;
        pkg.sign(signer)?;
        pkg.write_file(output)?;
    }

    println!("Signed: {}", output.display());

    Ok(())
}
